### PR TITLE
Retry s:delayed_checktime on 'E523: Not allowed'

### PR DESCRIPTION
**Pull request by Yilin-Yang**
Howdy!

On neovim, when using vim-tmux-focus-events with vim-airline, I've been seeing the error message:
```
  Error detected while processing function <SNR>52_delayed_checktime:
  line    8:
  E523: Not allowed here:   silent checktime
```
After performing the following steps:
1. Open neovim in tmux,
2. Type a command on the command line (e.g. `:asdfasdfasdf`) without
pressing enter,
3. Focus a different window (e.g. click on a Chrome window),
4. Click on the neovim/tmux terminal window again,
5. Press <ESC> to quit the command line.

[asciinema recording](https://asciinema.org/a/240886)

I'm using NVIM v0.4.0, but I've been seeing this message for a while, so it probably affects older versions of neovim as well. Standard vim doesn't seem to be affected.

This seems to occur despite the plugin already checking whether the user is in the command line (e.g. `mode() !=# 'c'` returns true in the affected function, even in calls where the error is thrown).

I tried patching it by silently catching the exception and letting the `focus_gained_checktime` fire a second time, after `mode()` has finished returning a correct value, which fixes the issue on my machine.

### CODE REVIEW

**Comment by bruno-** [comment]
Hi

> when using vim-tmux-focus-events with vim-airline

So, does the problem happen when `vim-airline` is not installed?

The proposed solution looks pretty hacky. Don't you think a better approach would be to try to improve `cursor_in_cmd_line` function?

**Comment by Yilin-Yang** [comment]
> So, does the problem happen when vim-airline is not installed?

As far as I can tell, this issue only occurs when using vim-tmux-focus-events, vim-airline, and neovim at the same time.

> would be to try to improve cursor_in_cmd_line function?

That's definitely a valid point, and it was actually the first thing I tried. The strange thing is that `s:cursor_in_cmd_line()` seems to be working correctly, as far as I can tell. I added [print statements to those functions](https://github.com/Yilin-Yang/vim-tmux-focus-events/blob/cursor_cmdline/autoload/tmux_focus_events.vim#L8-L31) to try and see what was happening, and got the following output:
<details>
<summary> neovim output </summary>

```
delayed_checktime called at: Sun 14 Apr 2019 08:49:58 PM EDT
cursor_in_cmd_line called at: Sun 14 Apr 2019 08:49:58 PM EDT
in_cmd_line: 1
in_cmd_window: 0
returning: 1
delayed_checktime: cursor_in_cmd_line returned true
in command line, exiting  " but evaluation of function continues?
Error detected while processing function <SNR>57_delayed_checktime:
line   11:
E523: Not allowed here:   silent checktime
delayed_checktime exited at: Sun 14 Apr 2019 08:49:59 PM EDT
```
</details>

It looks like the return statement in the `if <SID>cursor_in_cmd_line()` clause doesn't actually return from `s:delayed_checktime()`, for whatever reason.

In ordinary vim, this doesn't occur:
<details>
<summary> vim output </summary>

```
delayed_checktime called at: Sun 14 Apr 2019 08:56:46 PM EDT
cursor_in_cmd_line called at: Sun 14 Apr 2019 08:56:46 PM EDT
in_cmd_line: 1
in_cmd_window: 0
returning: 1
delayed_checktime: cursor_in_cmd_line returned true
in command line, exiting  " as expected
delayed_checktime called at: Sun 14 Apr 2019 08:56:48 PM EDT  " new function call
cursor_in_cmd_line called at: Sun 14 Apr 2019 08:56:48 PM EDT
in_cmd_line: 0
in_cmd_window: 0
returning: 0
delayed_checktime: cursor_in_cmd_line returned false
delayed_checktime exited at: Sun 14 Apr 2019 08:56:48 PM EDT " successfully 'checktime'd
```
</details>

**tl;dr** it looks like it might be an upstream issue with neovim, or at least just a difference between how vim and neovim execute VimL functions. I could definitely open an issue on the neovim repo (for a more permanent fix), but this might work as a quick stopgap. I'd understand if you weren't willing to merge it in lieu of a proper upstream fix, however.

**Comment by bruno-** [comment]
If we're going to go with this solution then maybe we don't need these 3 lines at the beginning of `s:delayed_checktime` function? Reason: we're now relying on `checktime` error and try catch logic for this.

```
if <SID>cursor_in_cmd_line()
  return
endif
```

**Comment by Yilin-Yang** [comment]
That's a good suggestion; the only reason I might hesitate to make that change is that the `catch /E523/` might not catch everything, i.e. there might be some errors other than `E523` that `silent checktime` can throw when the command line is open. One way around that would be to `catch` all thrown exceptions, but that might silence "real" error messages that we want the user to see.

I lean slightly towards leaving that if statement in `s:delayed_checktime`, but I don't have a strong preference either way. The code does also work with that if statement removed, in my own testing.

**Comment by bruno-** [comment]
> catch /E523/ might not catch everything

I hear you, I also wouldn't like to catch everything.

I also don't like to leave half-assed implementations out of the general fear "something may go wrong". I propose we take the stand and go with the `try..catch` solution only (remove the first 3 lines). If there are other errors, it's not the end of the world - users will report them and we'll handle them on the `catch` line. Thoughts?

**Comment by Yilin-Yang** [comment]
Works for me! I've made the change and pushed it to the remote branch. Were there any other things you wanted to address?

**Comment by bruno-** [comment]
Thanks!
